### PR TITLE
[bugfix] Accumulated_gradient and TensoBoard

### LINKED ISSUE #4304: TensorBoardLogger not working as expected with accumulate_grad_batches>1
## 🐛 Bug

When logging inside training step to TensorBoard and using `accumulate_grad_batches > 1` inside `pl.Trainer()` the behavior is not as expected.

With `accumulate_grad_batches = 1` everything looks good.
nice

With `accumulate_grad_batches = 8` the values are reported on the same step.

not_nice

### To Reproduce (sorry for not using colab)

```
import os

import torch
from torch.nn import functional as F
from torch.utils.data import DataLoader, random_split

import pytorch_lightning as pl
from pytorch_lightning.loggers import TensorBoardLogger
from torchvision.datasets.mnist import MNIST
from torchvision import transforms

class LitClassifier(pl.LightningModule):
    def __init__(self, hidden_dim=128, learning_rate=1e-3):
        super().__init__()
        self.save_hyperparameters()

        self.l1 = torch.nn.Linear(28 * 28, self.hparams.hidden_dim)
        self.l2 = torch.nn.Linear(self.hparams.hidden_dim, 10)

    def forward(self, x):
        x = x.view(x.size(0), -1)
        x = torch.relu(self.l1(x))
        x = torch.relu(self.l2(x))
        return x

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        self.log("train_loss",loss)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=self.hparams.learning_rate)

def run_test(accumulate_grad_batches,batch_size,num_workers):
    dataset = MNIST('', train=True, download=True, transform=transforms.ToTensor())
    mnist_train, mnist_val = random_split(dataset, [55000, 5000])
    train_loader = DataLoader(mnist_train,batch_size)
    val_loader = DataLoader(mnist_val,batch_size)

    model = LitClassifier()

    trainer = pl.Trainer(
        logger=TensorBoardLogger(os.getcwd()',name='bug'),
        accumulate_grad_batches=accumulate_grad_batches,
        max_epochs=2
        )
    trainer.fit(model, train_loader, val_loader)

run_test(1,32)
run_test(8,32)
```

### Expected behavior

Take a mean of the values logged at same step?

### Environment

 - PyTorch Version: 1.6
 - OS: Linux
 - How you installed PyTorch: conda
 - Python version: 3.8
 - Tensorboard: 2.3.0

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR tries to improve the logging display on TensorBoard when using accumulated_grad_batches > 1. 
It also introduces `log_epoch_metrics_on_step` parameter within the trainer.

`log_epoch_metrics_on_step` idea was dropped !

Fixes #4304

``log_epoch_metrics_on_step = True``

Screenshot 2020-11-18 at 11 38 55
Screenshot 2020-11-18 at 11 39 03

``log_epoch_metrics_on_step = False``

Screenshot 2020-11-18 at 11 57 42
Screenshot 2020-11-18 at 11 57 58

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 1]
```diff
@@ -181,7 +181,7 @@ def cache_training_step_metrics(self, opt_closure_result):
         self.logged_metrics.update(logged_metrics_tmp)
         self.cached_results.legacy_batch_log_metrics.update(logged_metrics_tmp)
 
-    def log_metrics(self, metrics, grad_norm_dic, step=None):
+    def log_metrics(self, metrics, grad_norm_dic, step=None, is_train_step=False):
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:184]: can we add docs what the`is_train_step` means?
```suggestion
    def log_metrics(self, metrics, grad_norm_dic, step=None, log_on_step=False):
```

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -271,7 +271,7 @@ def __init__(
                 sequence.
 
             val_check_interval: How often to check the validation set. Use float to check within a training epoch,
-                use int to check every n steps (batches).
+                use int åto check every n steps (batches).
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:274]:
```suggestion
                use to check every n steps (batches).
```

**Comment by edenlightning** [comment]
Ass discussed offline, i'd say this would be better as a param in log method than in the trainer.

**Comment by edenlightning** [comment]
@tchaton can we split this to 2 PRs?

**Comment by tchaton** [comment]
> @tchaton can we split this to 2 PRs?

Hey @edenlightning, I removed the parameters as you suggested. This PR contains only the fix for tensorboard logging in case of accumulated_gradient > 1

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 3]
```diff
@@ -95,14 +95,14 @@ def cache_logged_metrics(self) -> Union[EpochResultStore, None]:
...
             self._cached_results[self._current_stage].cache_result()
 
-    def on_trainer_init(self, logger, flush_logs_every_n_steps: int, log_every_n_steps: int, move_metrics_to_cpu: bool):
+    def on_trainer_init(self, logger, flush_logs_every_n_steps: int,
+                        log_every_n_steps: int, move_metrics_to_cpu: bool):
         # logging
         self.configure_logger(logger)
         # todo: IDE is complaining, these shall be initialized in the Trainer init at leas as placeholders
-        #  and assign here the desired value
+        # and assign here the desired value
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:103]: I think that the extra space is the to he parsed in IDE as single todo task

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 4]
```diff
@@ -181,7 +181,7 @@ def cache_training_step_metrics(self, opt_closure_result):
         self.logged_metrics.update(logged_metrics_tmp)
         self.cached_results.legacy_batch_log_metrics.update(logged_metrics_tmp)
 
-    def log_metrics(self, metrics, grad_norm_dic, step=None):
+    def log_metrics(self, metrics, grad_norm_dic, step=None, log_train_step_metrics=False):
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:184]: lets add it in docs and explain how it fuse with `step` argument
- what happen in `step=n` and `log_train_step_metrics=False`
- what happen in `step=None` and `log_train_step_metrics=True`

Comment (Borda) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:184]: shall it be `log_train_metrics_kind='step'|'epoch'`?

**Review by Borda — COMMENTED** [review]
sure about the APPI, and pls update docs

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:184]: Hey @jirka, 
`log_train_step_metrics` is there only for internal reason. 
This should not be exposed in the doc or made over-complicated :)
I wanted to have a simple boolean to keep things fast.

Comment (SkafteNicki) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:184]: could you add description of the new arg under docstring?

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:184]: done !

**Review by justusschock — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]
LGTM

**Review by SeanNaren — APPROVED** [review]

**Comment by tchaton** [comment]
Hey @williamFalcon , can you review this one ?

**Review by williamFalcon — APPROVED** [review]
